src/main/java/com/iexec/resultproxy/challenge/Eip712ChallengeService.java

```diff
@@ -41,15 +43,21 @@ Eip712Challenge generateEip712Challenge(Integer chainId) {
 
     private void saveEip712ChallengeString(String eip712ChallengeString) {
         challengeId++;
-        challengeMap.put(challengeId, eip712ChallengeString);
+        synchronized (challengeMap) {
```

Comment:
Why synchronizing since `ExpiringMap` is thread-safe?

Reply:
That's what I thought... But using `challengeMap.entrySet()` and iterating over the collection seems to break this thread-safety. That may be a bug of the library, I'm not sure.

Reply:
Is there any chance two `eip712Challenge` would be the same? If the response is negative, we could use them as keys for the expiring map instead of integer. We could then use `map.remove` method which is itself thread-safe.

Reply:
`eip712Challenge`s are supposed to be unique since they contain a randomly generated token. I think we can use them as keys and use dummy (or null) objects as values.

Reply:
I think the answer lies in the implementation of ExpiringMap:
https://github.com/jhalterman/expiringmap/blob/master/src/main/java/net/jodah/expiringmap/ExpiringMap.java#L689
To work in a thread-safe manner, we should only use methods implemented on the entrySet
This should forbid us to use `removeIf` as it does not seem to be supported, the reason should be because the production compiler is still Java 6
https://github.com/jhalterman/expiringmap/blob/expiringmap-0.5.10/pom.xml#L19

Reply:
Instead of using removeIf, we might copy map before iterating on it, in order to remove an element safely

Reply:
Thanks for all your inputs. The easiest solution seems to have the `eip712Challenge`s become the keys of the map. It seems to work perfectly. I'll push a commit once it is validated in local.

Reply:
:+1:
This should be much better given `eip712Challenge` are unique.
Moreover, challengeId was a simple `int` and not an `AtomicInteger` and I don't know what would have happened if 2 threads tried to save a challenge at the same time.